ludwig/models/ecd.py

```diff
@@ -69,6 +69,8 @@ def __init__(
         # ================ Combined loss metric ================
         self.eval_loss_metric = torchmetrics.MeanMetric()
 
+        self.to('cuda' if torch.cuda.is_available() else 'cpu')
```

Comment:
@jeffreyftang does removing this line cause failures?

I'm unsure about why this line is needed since we already move an initialized ECD object to to CUDA in the trainer [here](https://github.com/ludwig-ai/ludwig/pull/1508/files)

Reply:
Yes it does (although only 1 failure). 
```
>       return torch._C._nn.linear(input, weight, bias)
E       RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument mat1 in method wrapper_addmm)

/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/functional.py:1848: RuntimeError
```

Reply:
Originally I had modified `generic_decoders.py` to move all the decoders to GPU explicitly, but adding the single line to `ECD` resolved things (as well as cleared up some of the issues I encountered after making the mods to `generic_decoders.py`).

Reply:
Full error trace:
```
tests/integration_tests/utils.py:522: in run_api_experiment
    loaded_model.predict(dataset=data_csv)
ludwig/api.py:759: in predict
    dataset,
ludwig/models/predictor.py:118: in batch_predict
    preds = self._predict(model, batch)
ludwig/models/predictor.py:145: in _predict
    return model.predict_step(inputs)
ludwig/models/ecd.py:214: in predict_step
    return self.predictions(inputs, output_features=None)
ludwig/models/ecd.py:199: in predictions
    outputs = self(inputs)
/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py:1102: in _call_impl
    return forward_call(*input, **kwargs)
ludwig/models/ecd.py:159: in forward
    decoder_outputs = decoder(decoder_inputs, mask=mask)
/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py:1102: in _call_impl
    return forward_call(*input, **kwargs)
ludwig/features/base_feature.py:276: in forward
    logits = self.logits(logits_input, target=target)
ludwig/features/category_feature.py:191: in logits
    LOGITS: self.decoder_obj(hidden),
/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py:1102: in _call_impl
    return forward_call(*input, **kwargs)
ludwig/decoders/generic_decoders.py:166: in forward
    return self.dense(inputs)
/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py:1102: in _call_impl
    return forward_call(*input, **kwargs)
ludwig/utils/torch_utils.py:203: in forward
    output = torch.squeeze(self.dense(input), dim=-1)
/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py:1102: in _call_impl
    return forward_call(*input, **kwargs)
/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/modules/linear.py:103: in forward
    return F.linear(input, self.weight, self.bias)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

input = tensor([[ 0.6498, -3.4886, -3.3236, -2.5369, -1.2582,  0.6344,  3.4147, -1.1249],
        [ 1.1283, -0.5554,  1.0740, ....9143, -2.2151, -2.4031, -3.4755,  0.0326, -0.3848,  1.9021, -2.3640]],
       device='cuda:0', grad_fn=<SumBackward1>)
weight = Parameter containing:
tensor([[-0.3009,  0.5633,  0.1450, -0.3612, -0.3303, -0.0303,  0.3645, -0.1663],
        [-0.37...-0.3957],
        [-0.3995, -0.5854, -0.2124,  0.2221,  0.4184,  0.3013, -0.5777, -0.3927]],
       requires_grad=True)
bias = Parameter containing:
tensor([-0.0010,  0.0010,  0.0010,  0.0010, -0.0010,  0.0010],
       requires_grad=True)

    def linear(input: Tensor, weight: Tensor, bias: Optional[Tensor] = None) -> Tensor:
        r"""
        Applies a linear transformation to the incoming data: :math:`y = xA^T + b`.

        This operator supports :ref:`TensorFloat32<tf32_on_ampere>`.

        Shape:

            - Input: :math:`(N, *, in\_features)` N is the batch size, `*` means any number of
              additional dimensions
            - Weight: :math:`(out\_features, in\_features)`
            - Bias: :math:`(out\_features)`
            - Output: :math:`(N, *, out\_features)`
        """
        if has_torch_function_variadic(input, weight, bias):
            return handle_torch_function(linear, (input, weight, bias), input, weight, bias=bias)
>       return torch._C._nn.linear(input, weight, bias)
E       RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument mat1 in method wrapper_addmm)

/home/ray/anaconda3/lib/python3.7/site-packages/torch/nn/functional.py:1848: RuntimeError
========================================================================================================================================================================= short test summary info ==========================================================================================================================================================================
FAILED tests/integration_tests/test_api.py::test_api_intent_classification - RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argument mat1 in method wrapper_addmm)
```

Reply:
Jumping in without the full context, but could it be that the problem is with Dense, and how we initialize it? Like instead of `weights_initializer(self.dense.weight)` we should do `weights_initializer(self.dense.weight).to("coda")`? And the same for the bias?

Reply:
Discussed a bit with @ShreyaR offline, and I think the recommended approach is to move `ECD` to GPU from the predictor (similar to how `ECD` is moved in the trainer). Going to give it a try :)